### PR TITLE
BHBC-914: Save drafts

### PR DESCRIPTION
**Pull request by NickPhura**
<details open>
  <summary><a href="https://quartech.atlassian.net/browse/BHBC-914" title="BHBC-914" target="_blank">BHBC-914</a></summary>
  <br />
  <table>
    <tr>
      <th>Summary</th>
      <td>Implement the functionality to allow users to save a draft project</td>
    </tr>
    <tr>
      <th>Type</th>
      <td>
        Story
        Story
      </td>
    </tr>
    <tr>
      <th>Status</th>
      <td>In Progress</td>
    </tr>
    <tr>
      <th>Points</th>
      <td>N/A</td>
    </tr>
    <tr>
      <th>Labels</th>
      <td>-</td>
    </tr>
  </table>
</details>

---

# Overview

## Links to jira tickets

https://quartech.atlassian.net/browse/BHBC-914

## This PR contains the following changes

**APP**
Add create/update draft logic to create project page.

**API**
Generic draft endpoint to save a record to the webform_draft table.

**Misc**
Added the list of regions to the project view header (was a TODO before)
Altered the clamav container name in the docker-compose to match existing naming patterns

## This PR contains the following types of changes

- [x] New feature (change which adds functionality)
- [ ] Enhancement (improvements to existing functionality)
- [ ] Bug fix (change which fixes an issue)
- [x] Misc cleanup / Refactoring / Documentation
- [ ] Version change

## Checklist

A list of items that are good to consider when making any changes.

_Note: this list is not exhaustive, and not all items are always applicable._

### General

- [ ] I have performed a self-review of my own code

### Code

- [ ] New files/classes/functions have appropriately descriptive names and comment blocks to describe their use/behaviour
- [ ] I have avoided duplicating code when possible, moving re-usable pieces into functions
- [ ] I have avoided hard-coding values where possible and moved any re-usable constants to a constants file
- [ ] My code is as flat as possible (avoids deeply nested if/else blocks, promise chains, etc)
- [ ] My code changes account for null/undefined values and handle errors appropriately
- [ ] My code uses types/interfaces to help describe values/parameters/etc, help ensure type safety, and improve readability

### Style

- [ ] My code follows the established style conventions
- [ ] My code uses native material-ui components/icons/conventions when possible

### Documentation

- [ ] I have commented my code sufficiently, such that an unfamiliar developer could understand my code
- [ ] I have added/updated README's and related documentation, as needed

### Tests

- [ ] I have added/updated unit tests for any code I've added/updated
- [ ] I have added/updated the Postman requests/tests to account for any API endpoints I've added/updated

### Linting/Formatting

- [ ] I have run the linter and fixed any issues, as needed  
       _See the `lint` commands in package.json_
- [ ] I have run the formatter and fixed any issues, as needed  
       _See the `format` commands in package.json_

### SonarCloud

- [ ] I have addressed all SonarCloud Bugs, Vulnerabilities, Security Hotspots, and Code Smells

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Screenshots

Please add any relevant UI screenshots, if applicable.

### CODE REVIEW

**Review by anissa-agahchen — DISMISSED** [review]

File: api/src/database/db.ts [hunk 1]
```diff
@@ -39,11 +39,52 @@ try {
...
   commit: () => Promise<void>;
+  /**
+   * Rollsback the transaction, undoing any queries performed by this connection.
+   *
+   * Note: Does nothing if the connection is not open, or was released.
+   * @memberof IDBConnection
+   */
   rollback: () => Promise<void>;
+  /**
+   * Performs a query agaisnt this connection, returning the results.
```
Comment (sdevalapurkar) [inline api/src/database/db.ts]:
```suggestion
   * Performs a query against this connection, returning the results.
```

File: app/src/components/dialog/ErrorDialog.tsx [hunk 2]
```diff
@@ -100,13 +100,17 @@ export const ErrorDialog: React.FC<IErrorDialogProps> = (props) => {
...
-            <Button color="primary" onClick={() => setIsExpanded(!isExpanded)}>
-              {(isExpanded && 'Hide detailed error message') || 'Show detailed error message'}
-            </Button>
-            {props?.dialogErrorDetails?.length && (
-              <Collapse in={isExpanded}>
-                <ErrorDetailsList errors={props.dialogErrorDetails} />
-              </Collapse>
+            {props?.dialogErrorDetails?.length ? (
+              <>
+                <Button color="primary" onClick={() => setIsExpanded(!isExpanded)}>
```
Comment (sdevalapurkar) [inline app/src/components/dialog/ErrorDialog.tsx:105]: this is causing the errordialog.test.tsx test case to no longer be super accurate
can we change the message on the test to say:
```
renders correctly with a non-detailed error message
```

and then add a test case for with detailed error message :)

File: app/src/features/projects/CreateProjectPage.test.tsx [hunk 3]
```diff
@@ -213,4 +227,99 @@ describe('CreateProjectPage', () => {
...
+
+      await waitFor(() => {
+        expect(getByText('Save Incomplete Project as a Draft')).toBeVisible();
+      });
+
+      // Expect dialog to be open
+      expect(asFragment()).toMatchSnapshot();
+    });
+
+    it('the cancel button closes the dialog', async () => {
```
Comment (sdevalapurkar) [inline app/src/features/projects/CreateProjectPage.test.tsx]: supernit:
```suggestion
    it('closes the dialog on cancel button click', async () => {
```

File: app/src/features/projects/CreateProjectPage.test.tsx [hunk 4]
```diff
@@ -213,4 +227,99 @@ describe('CreateProjectPage', () => {
...
+
+      await waitFor(() => {
+        expect(queryByText('Save Incomplete Project as a Draft')).not.toBeInTheDocument();
+      });
+
+      // Expect dialog to be closed
+      expect(asFragment()).toMatchSnapshot();
+    });
+
+    it('the save button calls the createDraft/updateDraft functions and closes the dialog', async () => {
```
Comment (sdevalapurkar) [inline app/src/features/projects/CreateProjectPage.test.tsx]: supernit:
```suggestion
    it('calls the createDraft/updateDraft functions and closes the dialog on save button click', async () => {
```

Comment (sdevalapurkar) [reply app/src/features/projects/CreateProjectPage.test.tsx]: for readability for tests

File: app/src/features/projects/CreateProjectPage.tsx [hunk 5]
```diff
@@ -320,38 +333,61 @@ const CreateProjectPage: React.FC = () => {
...
+          dialogError: 'The response from the server was null, or did not contain a draft project ID.'
+        });
+
+        return;
+      }
+
+      setDraft({ id: response.id, date: response.date });
+    } catch (error) {
+      const apiError = error as APIError;
+      showErrorDialog({ dialogError: apiError?.message, dialogErrorDetails: apiError?.errors });
```
Comment (sdevalapurkar) [inline app/src/features/projects/CreateProjectPage.tsx]: is it possible to add a test case for this similar to other components?

File: app/src/features/projects/__snapshots__/CreateProjectPage.test.tsx.snap [hunk 6]
```diff
@@ -502,16 +5039,43 @@ exports[`CreateProjectPage renders the initial default page correctly 1`] = `
...
       <div
         class="MuiBox-root MuiBox-root-10"
+      >
+        <div
+          class="MuiBox-root MuiBox-root-11"
+        >
+          <span
+            class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorTextSecondary"
+          >
+            Draft saved on
```
Comment (sdevalapurkar) [inline app/src/features/projects/__snapshots__/CreateProjectPage.test.tsx.snap:6195]: question: there is no date here... should there be?

File: app/src/features/projects/components/ProjectDraftForm.tsx [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+export const ProjectDraftFormInitialValues: IProjectDraftForm = {
+  draft_name: ''
+};
+
+export const ProjectDraftFormYupSchema = yup.object().shape({
+  draft_name: yup.string().max(50, 'Cannot exceed 50 characters').required('Required')
+});
+
+/**
+ * Create project - General information section
```
Comment (sdevalapurkar) [inline app/src/features/projects/components/ProjectDraftForm.tsx]: looks like a copy paste out of date comment perhaps :)

File: app/src/features/projects/components/ProjectDraftForm.tsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+          <TextField
+            fullWidth
+            required={true}
+            id="draft_name"
+            name="draft_name"
+            label="Draft Name"
+            variant="outlined"
+            value={values.draft_name}
+            onChange={handleChange}
+            error={touched.draft_name && Boolean(errors.draft_name)}
```
Comment (sdevalapurkar) [inline app/src/features/projects/components/ProjectDraftForm.tsx:41]: can we please add a test case for this, just set up a test case with:

```
<Formik
        initialValues={projectDraftFilledValues}
        initialErrors={blah}
        initialTouched={blah}
        validationSchema={ProjectDraftFormYupSchema}
        validateOnBlur={true}
        validateOnChange={false}
        onSubmit={async (values) => {
          handleSaveAndNext(values);
        }}>
        {() => <ProjectDraftForm />}
</Formik>
```

File: app/src/features/projects/__snapshots__/CreateProjectPage.test.tsx.snap (hunk 6 above)
Comment (NickPhura) [reply app/src/features/projects/__snapshots__/CreateProjectPage.test.tsx.snap:6195]: Yeah its fine.  I render the string always, and just selectively make it visible.
Only reason is so the page doesnt get pushed down when the message shows up.  Probably a fancier way to do it, but I didnt want to get into it when this was a bonus thing I was adding not in the design.

**Review by sdevalapurkar — DISMISSED** [review]
fab work

**Comment by sdevalapurkar** [comment]
good to merge when CI passes (looks like test checks is failing currently)

**Comment by NickPhura** [comment]
> 
> 
> good to merge when CI passes (looks like test checks is failing currently)

Ya looks like a merge related issue. Easy fix.

**Review by sdevalapurkar — APPROVED** [review]

**Review by anissa-agahchen — COMMENTED** [review]
Looks good

**Pull request merged**